src/pam/main.cc

```diff
@@ -0,0 +1,422 @@
+  pid_t child_pid;
+
+  // Start the python subprocess
+  if (posix_spawnp(&child_pid, PYTHON_EXECUTABLE, nullptr, nullptr,
+                   const_cast<char *const *>(args), nullptr) != 0) {
+    syslog(LOG_ERR, "Can't spawn the howdy process: %s (%d)", strerror(errno),
+           errno);
+    return PAM_SYSTEM_ERR;
+  }
+
+  // NOTE: We should replace mutex and condition_variable by atomic wait, but
+  // it's too recent (C++20)
+  std::mutex m;
+  std::condition_variable cv;
+  ConfirmationType confirmation_type(ConfirmationType::Unset);
+
+  // This task wait for the status of the python subprocess (we don't want a
+  // zombie process)
+  optional_task<int> child_task([&] {
+    int status;
+    wait(&status);
+    {
+      std::unique_lock<std::mutex> lk(m);
+      if (confirmation_type == ConfirmationType::Unset) {
+        confirmation_type = ConfirmationType::Howdy;
+      }
+    }
+    cv.notify_one();
+
+    return status;
+  });
+  child_task.activate();
+
+  // This task waits for the password input (if the workaround wants it)
+  optional_task<std::tuple<int, char *>> pass_task([&] {
+    char *auth_tok_ptr = nullptr;
+    int pam_res = pam_get_authtok(
+        pamh, PAM_AUTHTOK, const_cast<const char **>(&auth_tok_ptr), nullptr);
+    {
+      std::unique_lock<std::mutex> lk(m);
+      if (confirmation_type == ConfirmationType::Unset) {
+        confirmation_type = ConfirmationType::Pam;
+      }
+    }
+    cv.notify_one();
+
+    return std::tuple<int, char *>(pam_res, auth_tok_ptr);
+  });
+
+  if (auth_tok) {
+    pass_task.activate();
+  }
+
+  // Wait for the end either of the child or the password input
+  {
+    std::unique_lock<std::mutex> lk(m);
+    cv.wait(lk, [&] { return confirmation_type != ConfirmationType::Unset; });
+  }
+
+  // The password has been entered or an error has occurred
+  if (confirmation_type == ConfirmationType::Pam) {
+    // We kill the child because we don't need its result
+    kill(child_pid, SIGTERM);
+    child_task.stop(false);
+
+    // We just wait for the thread to stop since it's this one which sent us the
+    // confirmation type
+    pass_task.stop(false);
+
+    char *password = nullptr;
+    std::tie(pam_res, password) = pass_task.get();
+
+    if (pam_res != PAM_SUCCESS) {
+      return pam_res;
+    }
+
+    // The password has been entered, we are passing it to PAM stack
+    return PAM_IGNORE;
+  }
+
+  // The compare process has finished its execution
+  child_task.stop(false);
+
+  // We want to stop the password prompt, either by canceling the thread when
+  // workaround is set to "native", or by emulating "Enter" input with
+  // "input"
+
+  // UNSAFE: We cancel the thread using pthread, pam_get_authtok seems to be
+  // a cancellation point
+  if (workaround == Workaround::Native && pass_task.is_active()) {
+    pass_task.stop(true);
+  } else if (workaround == Workaround::Input) {
+    if (geteuid() != 0) {
```

Comment:
A nitpick: The fake device _can_ be created by any user which has access to `/dev/uinput`. So forcing it to be root seems weird. Why not just stat the file to see if the current proc has access?

Reply:
That's because I didn't know about `euidaccess` which can test access for the effective UID (instead of the real one). It's fixed now.